### PR TITLE
Refactor source coordination to split interfaces between SourceCoordinator and SourceCoordinationStore

### PR DESCRIPTION
**Pull request by graytaylor0**
### Description
This change refactors the source coordination code, and adds some additional template and model classes for SourceCoordination

Changes:
* A `DefaultSourceCoordinator<T>` is implemented in data prepper core. This will be instantiated regardless of SourceCoordinationStore type, and will contain a `SourceCoordinationStore`. This will greatly eliminate duplicate code between stores, and will separate the `SourceCoordinator` responsbility from the `SourceCoordinationStore`
* SourcePartitionStatus enum to represent the status of a partition
* interface for `SourcePartitionItem` to be stored in the `SourceCoordinationStore`. This item is implemented for dynamodb through the `DynamoDbSourcePartitionItem`, which represents the ddb item schema.
* The `SourceCoordinationStore` interface is implemented through `DynamoDbSourceCoordinationStore`, which is now loaded as a plugin from the `SourceCoordinatorFactory`. This interface will be added to in the coming commits.
 
### Check List
- [x] New functionality includes testing.
- [x] New functionality has been documented.
  - [x] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStatusTest.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+import org.junit.Test;
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+
+public class SourcePartitionStatusTest {
+
+    @Test
+    public void testSourcePartitionStatus() {
+        assertThat(SourcePartitionStatus.ASSIGNED.name(), equalTo(SourcePartitionStatus.ASSIGNED.name()));
```
Comment (dlvenable) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStatusTest.java:17]: This appears to be a tautology. Why is this test here?

**Review by dlvenable — COMMENTED** [review]
Looks good. I have just a couple comments.

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStatusTest.java (hunk 1 above)
Comment (graytaylor0) [reply data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStatusTest.java:17]: Jacoco code coverage has to be 100%. I'll remove this once these are actually used in the future.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStoreItem.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
...
+ * @since 2.2
+ */
+public interface SourcePartitionStoreItem {
+    String getSourcePartitionKey();
+    String getPartitionOwner();
+    String getPartitionProgressState();
+    SourcePartitionStatus getSourcePartitionStatus();
+    Instant getPartitionOwnershipTimeout();
+    Instant getReOpenAt();
+    Long getClosedCount();
```
Comment (asifsmohammed) [inline data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStoreItem.java:22]: Question: What is the closed count used for?

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/SourcePartitionTest.java [hunk 3]
```diff
@@ -6,6 +6,8 @@
 package org.opensearch.dataprepper.model.source;
```
Comment (kkondaka) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/SourcePartitionTest.java:6]: I think this should be moved to `model/source/coordinator` directory? and change the package name accordingly.

File: data-prepper-plugins/dynamodb-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/dynamodb/DynamoDbSourcePartitionItem.java [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+import software.amazon.awssdk.enhanced.dynamodb.mapper.annotations.DynamoDbPartitionKey;
+
+import java.time.Instant;
+
+@DynamoDbBean
+public class DynamoDbSourcePartitionItem implements SourcePartitionStoreItem {
+
+    private String sourcePartitionKey;
+    private String partitionOwner;
+    private Long version;
```
Comment (asifsmohammed) [inline data-prepper-plugins/dynamodb-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/dynamodb/DynamoDbSourcePartitionItem.java:20]: Question: What is the version here? It's not part of the interface

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/SourcePartitionTest.java [hunk 5]
```diff
@@ -6,6 +6,8 @@
...
 
 import org.junit.jupiter.api.Test;
+import org.opensearch.dataprepper.model.source.coordinator.PartitionIdentifier;
+import org.opensearch.dataprepper.model.source.coordinator.SourcePartition;
```
Comment (asifsmohammed) [inline data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/SourcePartitionTest.java:10]: Why do you have these imports?

**Review by asifsmohammed — APPROVED** [review]

File: data-prepper-plugins/dynamodb-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/dynamodb/DynamoDbSourcePartitionItem.java (hunk 4 above)
Comment (graytaylor0) [reply data-prepper-plugins/dynamodb-source-coordination-store/src/main/java/org/opensearch/dataprepper/plugins/sourcecoordinator/dynamodb/DynamoDbSourcePartitionItem.java:20]: This version is a concept specific to DynamoDb, which is why it is not part of the interface, and we will use it for optimistic locking (https://docs.aws.amazon.com/amazondynamodb/latest/developerguide/DynamoDBMapper.OptimisticLocking.html)

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/SourcePartitionTest.java (hunk 5 above)
Comment (graytaylor0) [reply data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/SourcePartitionTest.java:10]: Not sure why these are here. I don't have them in my local branch. Will make sure they are gone on the next commit.

File: data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStoreItem.java (hunk 2 above)
Comment (graytaylor0) [reply data-prepper-api/src/main/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStoreItem.java:22]: The `closedCount` will be used to track the number of times that a partition has been closed. For example, if we are reading from an index in an OpenSearch cluster, we will read that index once, mark it as closed, and then at some point in the future (based on the `reOpenAt` timestamp), we will allow the partition to be processed again until we hit some configured max closedCount. The closedCount itself keeps us from infinitely pulling from the index (we need some way to mark the partition as COMPLETED)

**Review by kkondaka — APPROVED** [review]
Looks good.

**Review by kkondaka — APPROVED** [review]

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/SourcePartitionTest.java (hunk 3 above)
Comment (dlvenable) [reply data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/SourcePartitionTest.java]: Thanks for catching this!

File: data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStatusTest.java (hunk 1 above)
Comment (dlvenable) [reply data-prepper-api/src/test/java/org/opensearch/dataprepper/model/source/coordinator/SourcePartitionStatusTest.java:17]: Thanks for explaining. This works for now.

**Review by dlvenable — APPROVED** [review]

**Pull request merged**